File: src/II053.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def inorderSuccessor(self, root: 'TreeNode', p: 'TreeNode') -> 'TreeNode' | None:
        st: list[TreeNode] = []
        cur = root
        prev = None
        while st or cur:
            while cur:
                st.append(cur)
                cur = cur.left
            cur = st.pop()
            if prev == p:
                return cur
            prev = cur
            cur = cur.right
        return None

File: src/II053.cpp
#include <iostream>
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
  public:
  TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p) {
    TreeNode* ret = nullptr;
    while (root != nullptr) {
      if (root->val > p->val) {
        ret = root;
        root = root->left;
      } else {
        root = root->right;
      }
    }
    return ret;
  }
};
